File: Leetcode/PY/Leetcode1034-边界着色.py
class Solution:
    def colorBorder(self, grid: List[List[int]], row: int, col: int, color: int) -> List[List[int]]:
        n,m = len(grid),len(grid[0])
        st = [[0] * m for _ in range(n)]
        dx,dy = [0,1,0,-1],[1,0,-1,0]
        def dfs(x,y):
            is_bd = False
            for i in range(4):
                a,b = dx[i] + x,dy[i] + y
                if 0 <= a < n and 0 <= b < m and grid[a][b] == grid[x][y]:
                    if st[a][b] == 0 :
                        st[a][b] = 1
                        dfs(a,b)
                else:
                    is_bd = True

            if is_bd: st[x][y] = 2
        st[row][col] = 1
        dfs(row,col)
        for i in range(n):
            for j in range(m):
                if st[i][j] == 2: grid[i][j] = color
        return grid

File: Leetcode/JS/Leetcode1034-边界着色.js
var colorBorder = function(grid, row, col, color) {
    let n = grid.length,m = grid[0].length;
    let st = new Array(n).fill(0).map(() => new Array(m).fill(0));
    let dfs = function(x,y) {
        let is_bd = false;
        const dx = [0,1,0,-1],dy = [1,0,-1,0];
        for(let i = 0;i < 4;i++){
            let a = dx[i] + x,b = dy[i] + y;
            if(a >= 0 && a < n && b >= 0 && b < m && grid[x][y] == grid[a][b]) {
                if(st[a][b] == 0) {
                    st[a][b] = 1;
                    dfs(a,b);
                }
            }else{
                is_bd = true;
            }
        }
        if(is_bd) st[x][y] = 2;
    }
    st[row][col] = 1;
    dfs(row,col);
    for(let i = 0;i < n;i++){
        for(let j = 0;j < m;j++){
            if(st[i][j] == 2) grid[i][j] = color;
        }
    }
    return grid;
};